### PR TITLE
Require `BufRead` instead of `Read`

### PR DESCRIPTION
**Pull request by tcharding**
Require `BufRead` instead of `Read` for consensus decode trait.

### CODE REVIEW

**Comment by tcharding** [comment]
> Rebasing, undraftin, and eventually merging will unblock my work on push_decode.

Lets go!

**Comment by apoelstra** [comment]
Whew, this trait has weirdly-defined methods. I had to check the std docs to be sure that we were supposed to expose the internal buffer size like that with methods like "return the remaning contents of the internal buffer, unless there are none, in which case first fill it as much as we can".

utACK.

**Review by apoelstra — APPROVED** [review]
ACK 2a460011ef56dad345a544a4e1f93ad2babe5b64

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:120]: I believe the reason for this was that `Cursor` was intended to be compatible with the `Seek` trait. But we shouldn't need it and should be able to remove `Cursor` entirely.

File: bitcoin/src/blockdata/locktime/absolute.rs
Comment (Kixunil) [inline bitcoin/src/blockdata/locktime/absolute.rs:13]: I'd do the opposite: `use` the traits and remove `io::` everywhere.

**Review by Kixunil — COMMENTED** [review]
Holy shit, this change is horrible because of the blanket impl on `std` traits. I suggest removing the blanket impl instead. (The blanket impl makes other code horrible too.)

We can still have decent compatibility by implementing our traits for specific `std` types and providing bridging wrappers. Or just switch over to `lgio`.

File: bitcoin/src/blockdata/locktime/absolute.rs
Comment (tcharding) [reply bitcoin/src/blockdata/locktime/absolute.rs:13]: Can do.

**Review by Kixunil — DISMISSED** [review]
Please impl `BufRead` for byte slice and remove all those `Cursor` usages.

**Comment by Kixunil** [comment]
Also note that this is not an optimization but it prevents people from accidentally writing horribly unoptimal code.

**Comment by apoelstra** [comment]
> Holy shit, this change is horrible because of the blanket impl on `std` traits.

Without the blanket impl it's differently horrible (though maybe less horrible, on net). It means that we have a `BufRead` trait that isn't actually implemented on all the things that users expect `BufRead` to be implemented on. Yes, we can implement it specifically for `&[u8]` but then what happens when a user wants to use a file handle or something?

On second thought, in this case the blanket impl won't help them ... if they do not have `std` they won't get the blanket impl. So no matter what, their filehandle type needs to be aware of our `io` crate and its `BufRead` type. Ok, yeah, never mind. We should drop the blanket impl.

**Comment by apoelstra** [comment]
Since we already have `Cursor` I'd like to keep it for this PR. It is a very simple type and generally useful. Though I agree we don't need it on `&[u8]`.

I'm ambivalent about switching its internal position to `u64`. It looks like `std` uses `u64` (and to implement `BufRead` they need to cast the `usize` argument to a `u64` internally, which I suppose hypothetically could be a bug if it was used on a 128-bit arch and somebody was trying to consume enough bytes..). I believe you're correct that the motivation for this is the `Seek` trait, which maybe we will want in the future.

**Comment by Kixunil** [comment]
@apoelstra `&[u8]` already implements `Read` sanely (using `#[cfg]` trick), it really should implement `BufRead`. I personally never need `Cursor` in practice because of the impl and would hate to need it because someone didn't write a few lines of code.

When I wrote that comment I thought blanket impls are the problem and only later I discovered that it's solvable using `#[cfg]` and the solution is already being used. I no longer believe we should drop them.

File: io/src/lib.rs
Comment (tcharding) [reply io/src/lib.rs:120]: I'm going to drop this change an stick with the `u64`.

File: bitcoin/src/blockdata/locktime/absolute.rs
Comment (tcharding) [reply bitcoin/src/blockdata/locktime/absolute.rs:13]: Done in #2262, you asked for it, now you get to review it :)

File: io/src/lib.rs
Comment (tcharding) [reply io/src/lib.rs:117]: Sweet, thanks man.

File: io/src/lib.rs
Comment (tcharding) [inline io/src/lib.rs:290]: I don't know whats going on here, I can't get this to find the implementation of `BufRead`?
```bash
error[E0599]: no method named `fill_buff` found for array `[u8; 3]` in the current scope
   --> io/src/lib.rs:288:25
    |
288 |         let fill = data.fill_buff();
    |                         ^^^^^^^^^ method not found in `[u8; 3]`
```

And explicitly using a reference to the 
```rust
        fn slice(data: &[u8]) {
            let fill = data.fill_buff();
        }
        let data = &[0_u8, 1, 3];
        slice(data);
```
```bash
error[E0599]: no method named `fill_buff` found for reference `&[u8]` in the current scope
   --> io/src/lib.rs:288:25
    |
288 |         let fill = data.fill_buff();
    |                         ^^^^^^^^^ method not found in `&[u8]`
```

Comment (Kixunil) [reply io/src/lib.rs:290]: It's one `f`. `fill_buf` not `fill_buff`. Also the reason first one doesn't work is you need `let mut fill = &[0, 1, 3] as &[_];` so that `.fill_buf` can make `&mut &[u8]` from it.

Comment (tcharding) [reply io/src/lib.rs:290]: Oh my goodness - face palm. I spent like 30 minutes on that.

Comment (Kixunil) [reply io/src/lib.rs:290]: Yeah, something like that happened to me. It seems like we rely on rustc so much that in those few occasions when it doesn't suggest a typo was made we don't consider it.

**Comment by tcharding** [comment]
Just to be clear, this PR was originally @Kixunil's work in https://github.com/rust-bitcoin/rust-bitcoin/pull/2179

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:96]: I think a (debug) panic would be better. Passing large value is considered invalid anyway.

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:116]: How do we know it's not recursion? Use UFCS to disambiguate?

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:183]: Why not just compare `self.inner.as_ref().len()`? Sure if the callers do something stupid like implementing `AsRef<[u8]>` to return different slices they will get problems but that applies anyway.

**Review by Kixunil — COMMENTED** [review]
A few things look too sus unfortunately.

File: io/src/lib.rs
Comment (tcharding) [reply io/src/lib.rs:116]: I tried both originally and thought the compiler must be able to figure it out. Anyways, used UFCS now.

File: io/src/lib.rs
Comment (tcharding) [reply io/src/lib.rs:96]: I used `assert!` since the stdlib version panics if amount is too big - added a panics section to the  rustdocs of the trait also.

File: io/src/lib.rs
Comment (tcharding) [reply io/src/lib.rs:183]: Used an `assert!` instead.

File: io/src/lib.rs
Comment (apoelstra) [inline io/src/lib.rs:114]: In b4696c449633e2078a79f71df3a9e75782afea89:

I see we already do this for `Read`, so I don't think we should hold up this PR for it, but does this code interact badly with the coherency rules? I think it means that users can implement this trait for their own types only if `std` is off ... and then if somebody turns `std` on, their code will break.

**Review by apoelstra — APPROVED** [review]
ACK 15faa20fd45a0656c34c534edaf394307db93f00

File: io/src/lib.rs
Comment (Kixunil) [reply io/src/lib.rs]: > added a panics section to the rustdocs of the trait also.

I'm not sure it's needed since it's already documented to be invalid in the trait itself but doesn't hurt.

File: io/src/lib.rs
Comment (Kixunil) [reply io/src/lib.rs:114]: I think the only way for it to happen is to depend on `std` and simultaneously on `bitcoin-io` without `std`. I think that shouldn't happen since activating `std` in a downstream crate ought to activate `std` here and in every other crate.

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:93]: I know why it's written like this but could have a comment about it so someone doesn't come up with a "bright" idea of changing this to `cmp::min(buf.len(), self.remaining as usize)`.

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:146]: Huh, why is this variable still here?

File: io/src/lib.rs
Comment (apoelstra) [reply io/src/lib.rs:114]: Yeah, I think you're right.

File: io/src/lib.rs
Comment (tcharding) [reply io/src/lib.rs:146]: Yes, cleaner without. Thanks.

**Comment by tcharding** [comment]
Draft until I resolve @Kixunil's review comments.

File: io/src/lib.rs
Comment (tcharding) [reply io/src/lib.rs:93]: Alright, you are going to have to help me out here. I've looked at this a couple of separate times now and I can't find the case where the following two lines create a different `cap` value.
```rust
let cap = cmp::min(buf.len() as u64, self.remaining) as usize;
let cap = cmp::min(buf.len(), self.remaining as usize);
```

Comment (Kixunil) [reply io/src/lib.rs:93]: The latter may overflow on 32 bit architecture.

Comment (tcharding) [reply io/src/lib.rs:93]: Ahh, because `Read::take` accepts a `u64`. Thanks.

Comment (tcharding) [reply io/src/lib.rs:95]: Added the following comment
```rust
        // Cast length to a u64 instead of casting `remaining` to a `usize`
        // (in case `remaining > u32::MAX` and we are on a 32 bit machine).
        let cap = cmp::min(buf.len() as u64, self.remaining) as usize;
```

**Review by Kixunil — CHANGES_REQUESTED** [review]
Aside from the bug, you previously set the PR to draft until you address my feedback but you now un-drafted it without removing `remaining` from `Cursor`...

File: io/src/lib.rs
Comment (tcharding) [reply io/src/lib.rs:101]: Bollocks, same as you said in the other place.

**Comment by tcharding** [comment]
> Aside from the bug, you previously set the PR to draft until you address my feedback but you now un-drafted it without removing `remaining` from `Cursor`...

Ha, that is a comical fail for the start of the year. I literally wondered yesterday about `remaining` after it appeared to come back (I must have swapped machines and not pulled), I convinced myself it was ok - face palm.

**Comment by tcharding** [comment]
Force push removes `remaining` ... again. Also casts `usize` to `u64` instead of the other way around.

**Review by apoelstra — APPROVED** [review]
ACK 51f519e81f20e1307dd7813142a158b0ddf8ab68

**Comment by sanket1729** [comment]
> Also note that this is not an optimization but it prevents people from accidentally writing horribly unoptimal code.

Is it that bad?

**Review by sanket1729 — APPROVED** [review]
ACK 51f519e81f20e1307dd7813142a158b0ddf8ab68. We should still wait for Kix's ACK

**Comment by tcharding** [comment]
Thanks for the review @sanket1729, and yes please wait for @Kixunil before merging.

**Review by Kixunil — CHANGES_REQUESTED** [review]
Breaks my heart to find a bug in two-ACK PR. :(

File: bitcoin/src/consensus/serde.rs
Comment (apoelstra) [reply bitcoin/src/consensus/serde.rs:455]: Oof, yeah.

I think it'd be clearest to change
```
if let Some(first) = self.buf
```
to
```
if let Some(first) = self.buf.take()
```
Which would also visually flag all the other `if let Some(first) = self.buf` instances so reviewers would think "should this be a `take`?".

Comment (apoelstra) [reply bitcoin/src/consensus/serde.rs:455]: Ah, I see. I doubt the perf difference is visible. Plausibly it could even improve performance by making branch prediction easier.

Comment (Kixunil) [reply bitcoin/src/consensus/serde.rs:455]: You copied the bug from my buggy code so it's primarily my fault.

File: io/src/lib.rs
Comment (tcharding) [reply io/src/lib.rs:55]: I don't remember writing that, I've removed it and better described the PR. Thanks, attention to details like this is one of the reasons I like working in this project. Please always pick me up on anomalies like this, I try to write descriptive _correct_ commit logs that match the change because I think it makes for better patch sets and better reviews.

**Comment by tcharding** [comment]
Changes in force push:
- Used `take` .
- Updated the commit log in the first patch.

File: io/src/lib.rs
Comment (Kixunil) [reply io/src/lib.rs:55]: I believe you wrote it before I told you how to fix the compilation issue - the CI was red at the time.

And if not for the bug I wouldn't like to block two-acked PR for something like this. Especially not *this PR*.

File: bitcoin/src/consensus/serde.rs
Comment (Kixunil) [inline bitcoin/src/consensus/serde.rs:452]: You forgot to move this out.

**Review by Kixunil — CHANGES_REQUESTED** [review]

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:60]: Since you have to modify the PR anyway, I suggest changing this to "Marks the buffered data up to `amount` as consumed" since the method actually doesn't consume anything.

**Comment by Kixunil** [comment]
@sanket1729 

> Is it that bad?

Yes. Storytime!

Back when I was a kid, I've implemented my own encryption program (yes, the horror; please ignore obvious cryptographic issues; to my defense though I wasn't defending against state-level attackers...) It took a file to encrypt and a key file which was just random noise it xorred each byte of the input file with a corresponding byte of the output file - reading and writing one-by-one.

What would one expect to take under a second took minutes (IIRC, anyway, it was clearly crazy slow) for even few hundred kB files. And as a kid I didn't know about cost of context switch and how buffering avoids it.

We don't parse byte-by-byte but there's a lot of places with similarly short reads and the only thing that actually does bulk reads are scripts and witness elements. So yes, I expect it'd be horrible to make that mistake and could be hard for newbies. Notably I still see questions on /r/rust "why is reading from file slow in Rust?" from time to time.

File: bitcoin/src/consensus/serde.rs
Comment (tcharding) [reply bitcoin/src/consensus/serde.rs:452]: Holly shit, might have been quicker to do this PR yourself. Sorry bro, sloppy work.

File: io/src/lib.rs
Comment (tcharding) [reply io/src/lib.rs:60]: Used as suggested.

**Comment by tcharding** [comment]
Changes in force push:
- Fixed the comment.
- Removed the _new_ bug I just introduced this morning.

File: bitcoin/src/consensus/serde.rs
Comment (Kixunil) [inline bitcoin/src/consensus/serde.rs:449]: Looks like you've removed the if instead of moving it...

**Review by Kixunil — CHANGES_REQUESTED** [review]

**Comment by Kixunil** [comment]
Took me a while to understand your reference to "morning". It's midnight here. :laughing:

File: bitcoin/src/consensus/serde.rs
Comment (tcharding) [reply bitcoin/src/consensus/serde.rs:449]: I'm not sure what you were expecting but ..

We have the `BufRead` implementation reading into a single byte buffer and the `Read` implementation reading as much of the iterator as will fit in the `buf` argument. That seems surprising. Wouldn't it be better to use some arbitrary buffer 1024 bytes or some thing? 

(My fail again for not thinking hard enough during when doing the original PR.)

Comment (apoelstra) [reply bitcoin/src/consensus/serde.rs:449]: The `IterReader` only reads and buffers a single byte at a time because that's what we get from an iterator. I agree it's a weird API and I'd need to reread the PR to remember what's going on here.

But that has nothing to do with the code at hand, which used to check whether `buf` was empty (and in this case it would do nothing), and now panics instead.

Comment (apoelstra) [reply bitcoin/src/consensus/serde.rs:449]: The `IterReader` only reads and buffers a single byte at a time because that's what we get from an iterator. I agree it's a weird API and I'd need to reread the PR to remember what's going on here.

But that has nothing to do with the code at hand, which used to check whether `buf` was empty (and in this case it would do nothing), and now panics instead.

**Comment by tcharding** [comment]
Ok, I'm seriously having a dismal day today. I finally worked it out - I think. Please confirm that the usage of `buf[1..]` then call to `iter` is ok. I believe Rust handles this even if `buf` only had one element to start with but after the slice (and after todays efforts I'm not totally sure).

File: bitcoin/src/consensus/serde.rs
Comment (Kixunil) [reply bitcoin/src/consensus/serde.rs:453]: Exactly, the issue is panic when an empty slice is provided (the type really should've been `NonEmptyOutBuf` but it's too late for that).

Regarding buffer length, since the input is already buffered and the hex decoding doesn't involve context switch or dynamic dispatch I assume it wouldn't help performance but quite likely make it worse. Maybe unless it prevents SIMD or something. If anyone has a benchmark proving otherwise I'd be happy to accept a change but I don't consider it a priority to write it myself.

**Review by Kixunil — APPROVED** [review]
ACK 263a8b3603cffd8e37ab6f54fc14d46ff9fb85d0

**Comment by yancyribbens** [comment]
> Back when I was a kid, I've implemented my own encryption program

The first rule of crypto is don't implement your own crypto because you probably won't get it right :P

**Review by apoelstra — APPROVED** [review]
ACK 263a8b3603cffd8e37ab6f54fc14d46ff9fb85d0

**Comment by Kixunil** [comment]
@stevenroose how about giving an example of a perfectly reasonable code that breaks with this change? And FTR the age old API will break anyway as we change it to be push-decoding so I don't think it's a big deal to also do this.

**Comment by stevenroose** [comment]
> @stevenroose how about giving an example of a perfectly reasonable code that breaks with this change? And FTR the age old API will break anyway as we change it to be push-decoding so I don't think it's a big deal to also do this.

push-decoding?

**Comment by Kixunil** [comment]
See #2184

**Pull request merged**
